library/list.lisp

```diff
@@ -283,31 +292,33 @@
        (if (== x e)
            True
            (member e xs)))
-      ((Nil)
-       False)))
+      ((Nil) False)))
 
   (declare union (Eq :a => ((List :a) -> (List :a) -> (List :a))))
   (define (union xs ys)
     "Returns a new list with the elements from both XS and YS and without duplicates."
-    (match xs
-      ((Cons x xs)
-       (if (or (member x ys)
-               (member x xs))
-           (union xs ys)
-           (Cons x (union xs ys))))
-      ((Nil) (remove-duplicates ys))))
+    (let ((rec
```

Comment:
anyone know why our `union` doesn't just call `cl:union`?

Reply:
is it a pain to call the type class method `==` in the `:test`?

Reply:
a little bit; you need to use a helper function that returns a tuple of `(:a, (:a -> :a -> Boolean))` to build a closure over the instance. `%get-hash-test-funcs` does that in hashtable.lisp, and it's a little nasty. i think scoped-type-variables would eliminate the need for that.

Reply:
Another caveat with this is that SBCL's `union` uses a hash-table under the hood, and we can't guarantee that `Hash` will ever be implemented.

Reply:
looking at the source for SBCL's `union`, it looks like they only use a hash-table when the test is one of the standard hash-table tests (`eq`, `eql`, `equal`, `equalp`), and otherwise do a relatively naïve `remove-duplicates` that's essentially the same as ours. our `==` will never be one of the standard hash-table tests, so at this point i don't see any reason to call `cl:union`.

that said, i think it would be reasonable to require `Hash :a` and remove duplicates by hash table; that's what i wrote for the iterator version.

eh.